--- linux-3.5.6/drivers/char/ppdev.c	2012-10-07 23:36:20.000000000 +0800
+++ linux-3.8.13/drivers/char/ppdev.c	2013-05-12 04:57:46.000000000 +0800
@@ -251,12 +251,8 @@
 			break;
 		}
 
-		if (signal_pending (current)) {
-			if (!bytes_written) {
-				bytes_written = -EINTR;
-			}
+		if (signal_pending (current))
 			break;
-		}
 
 		cond_resched();
 	}
@@ -783,7 +779,8 @@
 		err = PTR_ERR(ppdev_class);
 		goto out_chrdev;
 	}
-	if (parport_register_driver(&pp_driver)) {
+	err = parport_register_driver(&pp_driver);
+	if (err < 0) {
 		printk (KERN_WARNING CHRDEV ": unable to register with parport\n");
 		goto out_class;
 	}
